### PR TITLE
Fix a bug in olm install

### PR DESCRIPTION
**Pull request by nunnatsa**
When installing OLM, retry creating resources, if the CRD is not ready yet.

Fixes: #6489

### Description of the change
On olm install command, if the creation of a resource fails with "no-match" error, retry creating it, assuming the CRD is not ready yet, but is going to be ready very soon.

### Motivation for the change
Bug fix. See #6489 

##Checklist

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by nunnatsa** [comment]
Result of running the `olm install` command in v1.30.0:
``` bash
$ operator-sdk olm install --verbose
DEBU[0000] Debug logging is set                         
INFO[0000] Fetching CRDs for version "latest"           
INFO[0000] Fetching resources for resolved version "latest" 
INFO[0001] Creating CRDs and resources                  
INFO[0001]   Creating CustomResourceDefinition "catalogsources.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "clusterserviceversions.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "installplans.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "olmconfigs.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "operatorconditions.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "operatorgroups.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "operators.operators.coreos.com" 
INFO[0001]   Creating CustomResourceDefinition "subscriptions.operators.coreos.com" 
INFO[0001]   Creating Namespace "olm"                   
INFO[0001]   Creating Namespace "operators"             
INFO[0001]   Creating ServiceAccount "olm/olm-operator-serviceaccount" 
INFO[0001]   Creating ClusterRole "system:controller:operator-lifecycle-manager" 
INFO[0001]   Creating ClusterRoleBinding "olm-operator-binding-olm" 
INFO[0001]   Creating OLMConfig "cluster"               
FATA[0001] Failed to install OLM version "latest": failed to create CRDs and resources: no matches for kind "OLMConfig" in version "operators.coreos.com/v1" 
```

Result after this fix; notice the `OLMConfig` resource:
``` bash
$ build/operator-sdk olm install                   
INFO[0000] Fetching CRDs for version "latest"           
INFO[0000] Fetching resources for resolved version "latest" 
INFO[0001] Creating CRDs and resources                  
INFO[0001]   Creating CustomResourceDefinition "catalogsources.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "catalogsources.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "clusterserviceversions.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "clusterserviceversions.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "installplans.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "installplans.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "olmconfigs.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "olmconfigs.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "operatorconditions.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "operatorconditions.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "operatorgroups.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "operatorgroups.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "operators.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "operators.operators.coreos.com" created 
INFO[0001]   Creating CustomResourceDefinition "subscriptions.operators.coreos.com" 
INFO[0001]     CustomResourceDefinition "subscriptions.operators.coreos.com" created 
INFO[0001]   Creating Namespace "olm"                   
INFO[0001]     Namespace "olm" created                  
INFO[0001]   Creating Namespace "operators"             
INFO[0001]     Namespace "operators" created            
INFO[0001]   Creating ServiceAccount "olm/olm-operator-serviceaccount" 
INFO[0001]     ServiceAccount "olm/olm-operator-serviceaccount" created 
INFO[0001]   Creating ClusterRole "system:controller:operator-lifecycle-manager" 
INFO[0001]     ClusterRole "system:controller:operator-lifecycle-manager" created 
INFO[0001]   Creating ClusterRoleBinding "olm-operator-binding-olm" 
INFO[0001]     ClusterRoleBinding "olm-operator-binding-olm" created 
INFO[0001]   Creating OLMConfig "cluster"               
INFO[0001]     the CRD for OLMConfig "cluster" is not ready yet. Retrying in one second 
INFO[0004]     OLMConfig "cluster" created              
INFO[0004]   Creating Deployment "olm/olm-operator"     
INFO[0004]     Deployment "olm/olm-operator" created    
INFO[0004]   Creating Deployment "olm/catalog-operator" 
INFO[0004]     Deployment "olm/catalog-operator" created 
INFO[0004]   Creating ClusterRole "aggregate-olm-edit"  
INFO[0004]     ClusterRole "aggregate-olm-edit" created 
INFO[0004]   Creating ClusterRole "aggregate-olm-view"  
INFO[0004]     ClusterRole "aggregate-olm-view" created 
INFO[0004]   Creating OperatorGroup "operators/global-operators" 
INFO[0004]     OperatorGroup "operators/global-operators" created 
INFO[0004]   Creating OperatorGroup "olm/olm-operators" 
INFO[0004]     OperatorGroup "olm/olm-operators" created 
INFO[0004]   Creating ClusterServiceVersion "olm/packageserver" 
INFO[0004]     ClusterServiceVersion "olm/packageserver" created 
INFO[0004]   Creating CatalogSource "olm/operatorhubio-catalog" 
INFO[0004]     CatalogSource "olm/operatorhubio-catalog" created 
INFO[0004] Waiting for deployment/olm-operator rollout to complete 
INFO[0004]   Waiting for Deployment "olm/olm-operator" to rollout: 0 of 1 updated replicas are available 
INFO[0005]   Deployment "olm/olm-operator" successfully rolled out 
INFO[0005] Waiting for deployment/catalog-operator rollout to complete 
INFO[0005]   Deployment "olm/catalog-operator" successfully rolled out 
INFO[0005] Waiting for deployment/packageserver rollout to complete 
INFO[0005]   Waiting for Deployment "olm/packageserver" to rollout: 0 of 2 updated replicas are available 
INFO[0012]   Deployment "olm/packageserver" successfully rolled out 
INFO[0012] Successfully installed OLM version "latest"  

NAME                                            NAMESPACE    KIND                        STATUS
catalogsources.operators.coreos.com                          CustomResourceDefinition    Installed
clusterserviceversions.operators.coreos.com                  CustomResourceDefinition    Installed
installplans.operators.coreos.com                            CustomResourceDefinition    Installed
olmconfigs.operators.coreos.com                              CustomResourceDefinition    Installed
operatorconditions.operators.coreos.com                      CustomResourceDefinition    Installed
operatorgroups.operators.coreos.com                          CustomResourceDefinition    Installed
operators.operators.coreos.com                               CustomResourceDefinition    Installed
subscriptions.operators.coreos.com                           CustomResourceDefinition    Installed
olm                                                          Namespace                   Installed
operators                                                    Namespace                   Installed
olm-operator-serviceaccount                     olm          ServiceAccount              Installed
system:controller:operator-lifecycle-manager                 ClusterRole                 Installed
olm-operator-binding-olm                                     ClusterRoleBinding          Installed
cluster                                                      OLMConfig                   Installed
olm-operator                                    olm          Deployment                  Installed
catalog-operator                                olm          Deployment                  Installed
aggregate-olm-edit                                           ClusterRole                 Installed
aggregate-olm-view                                           ClusterRole                 Installed
global-operators                                operators    OperatorGroup               Installed
olm-operators                                   olm          OperatorGroup               Installed
packageserver                                   olm          ClusterServiceVersion       Installed
operatorhubio-catalog                           olm          CatalogSource               Installed
```

File: internal/olm/client/client.go [hunk 1]
```diff
@@ -113,17 +113,44 @@ func (c Client) DoCreate(ctx context.Context, objs ...client.Object) error {
 	for _, obj := range objs {
 		kind := obj.GetObjectKind().GroupVersionKind().Kind
 		log.Infof("  Creating %s %q", kind, getName(obj.GetNamespace(), obj.GetName()))
-		err := c.KubeClient.Create(ctx, obj)
-		if err != nil {
-			if !apierrors.IsAlreadyExists(err) {
+
+		for i := 0; i < 5; i++ { // try to create 5 times before giving up
```
Comment (oceanc80) [inline internal/olm/client/client.go:117]: It looks like if we get the `NoMatchError` 5 times, we'll swallow the error and return nil. We probably would want to surface the error in this case instead of swallowing it.

**Comment by oceanc80** [comment]
We probably want a thorough set of unit tests to both reproduce the original error and exercise the edge cases of the new retry loop.

File: internal/olm/client/client.go [hunk 2]
```diff
@@ -113,17 +113,44 @@ func (c Client) DoCreate(ctx context.Context, objs ...client.Object) error {
...
 
+func (c Client) tryCreatingOneResource(ctx context.Context, obj client.Object, kind string) (bool, error) {
+	err := c.KubeClient.Create(ctx, obj)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			log.Infof("    %s %q already exists", kind, getName(obj.GetNamespace(), obj.GetName()))
+			return false, nil
+		}
+
+		if meta.IsNoMatchError(err) {
```
Comment (oceanc80) [inline internal/olm/client/client.go:144]: nit: could use an `if/else if/else` statement here to more clearly semantically signal that the cases being checked for are related.

File: internal/olm/client/client.go (hunk 1 above)
Comment (nunnatsa) [reply internal/olm/client/client.go:117]: :+1: good point. But in that case, what do you say about making it a lop of ten, just for case?

File: internal/olm/client/client.go (hunk 2 above)
Comment (nunnatsa) [reply internal/olm/client/client.go:144]: I guess it's a matter of style. I'm not sure that go-simple linter will agree here, for instance (`else` after `return`)

File: internal/olm/client/client.go [hunk 3]
```diff
@@ -112,16 +112,52 @@ func NewClientForConfig(cfg *rest.Config) (*Client, error) {
...
+
+		if !retry {
+			log.Infof("    %s %q created", kind, resourceName)
+			return nil
+		}
+
+		log.Infof("    the CRD for %s %q is not ready yet. Retrying in one second", kind, resourceName)
+		time.Sleep(time.Second)
+	}
+
```
Comment (joelanford) [inline internal/olm/client/client.go:142]: It would be a bit more idiomatic (at least from our kubernetes adjacency) to use the client-go retry package.

See: https://pkg.go.dev/k8s.io/client-go/util/retry#OnError

File: internal/olm/client/client.go [hunk 4]
```diff
@@ -112,16 +112,52 @@ func NewClientForConfig(cfg *rest.Config) (*Client, error) {
...
+		log.Infof("  Creating %s %q", kind, resourceName)
+
+		if err := c.safeCreateOneResource(ctx, obj, kind, resourceName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// try to create 10 times before giving up
```
Comment (grokspawn) [inline internal/olm/client/client.go:126]: Is there a reason for the 10?  I was going to ask about the 5, but since then it's become 10, and I wonder if it might indeed become something else in future, since we're dealing with unknown forces which cause an atomic step to fail.  
I like that you're solving the general case instead of the specific one, but would it be possible for you to define an environment variable which would be used for this value, and if not found would default to 10? 

Something like 
```golang
const resourceCreateRetryCountEnv = "SDK_RESOURCE_RETRY_COUNT"

func getResourceCreateRetryCount() int {
   if s := os.Getenv(resourceCreateRetryCountEnv); s != "" {
     if i, e := strconv.Atoi(s); e == nil {
       return i
     }
   }
   
   return 10
}
```

and then the for loop is 
```golang
max = getResourceCreateRetryCount()
for i:=0; i < max; i++ {
```

**Review by everettraven — COMMENTED** [review]
Thanks for the PR @nunnatsa ! 

I have a general question - does retrying actually resolve the error?

 I am wondering if part of the problem is that there is no logic in the installation method in https://github.com/operator-framework/operator-sdk/blob/861f524d4731cd2d6bdbead0f071530fb07fe72b/internal/olm/installer/client.go#L88-L91 that differentiates between the actual `CustomResourceDefinition` resources and the instances of the CRDs so the ordering in which these resources are being created could be out of whack. This makes me think that retrying won't suffice in the event the CRD hasn't even been created yet due to the ordering of resource creation. I am thinking it might make more sense to have a distinct step for creating the CRDs that verifies all the CRDs have been created and then attempts to create the resources. (although it does appear to be the case that the list of resources is sorted such that CRDs come first here: https://github.com/operator-framework/operator-sdk/blob/861f524d4731cd2d6bdbead0f071530fb07fe72b/internal/olm/installer/client.go#L210 )
 
 Also leaving a couple comments:

**Comment by nunnatsa** [comment]
Thank you all. You raised some very good points, and I was trying to address them all.

The `timeout` flag is not available for this function, but it's used to create the context. The `retry.OnError` func is cool, but it does not work well with timeout context, at least in my tests here. I eventually used the `wait.ExponentialBackoffWithContext` function that is very similar, but does check the context. It requires some more logic code, but not too much.

As for the ordering: this is a great point and I checked it before I even started the implementation. For now, all the CRDs are deployed before the resources. I fully agree that for the long run, it will be better to split the CRD creation and the resource creation. But this refactoring is out of my capacity right now - sorry.

Anyway, since the order is right, then this solution solves the problem. In my environment I **always** see (only) one failure message, and then the creation succeeds.

File: internal/olm/client/client.go [hunk 5]
```diff
@@ -112,18 +111,50 @@ func NewClientForConfig(cfg *rest.Config) (*Client, error) {
...
+
+		if meta.IsNoMatchError(err) {
+			log.Infof("    Failed to create %s %q. CRD is not ready yet. Retrying...", kind, resourceName)
+			return false, nil
+		}
+
+		return false, err
+	})
+
+	if err != nil && err == context.DeadlineExceeded {
```
Comment (everettraven) [inline internal/olm/client/client.go:151]: Nit:
```suggestion
	if err != nil && errors.Is(err, context.DeadlineExceeded) {
```

File: changelog/fragments/01-olm-install-retry.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      INFO[0001]   Creating CustomResourceDefinition "operatorgroups.operators.coreos.com" 
+      INFO[0001]   Creating CustomResourceDefinition "operators.operators.coreos.com" 
+      INFO[0001]   Creating CustomResourceDefinition "subscriptions.operators.coreos.com" 
+      INFO[0001]   Creating Namespace "olm"                   
+      INFO[0001]   Creating Namespace "operators"             
+      INFO[0001]   Creating ServiceAccount "olm/olm-operator-serviceaccount" 
+      INFO[0001]   Creating ClusterRole "system:controller:operator-lifecycle-manager" 
+      INFO[0001]   Creating ClusterRoleBinding "olm-operator-binding-olm" 
+      INFO[0001]   Creating OLMConfig "cluster"               
+      FATA[0001] Failed to install OLM version "latest": failed to create CRDs and resources: no matches for kind "OLMConfig" in version "operators.coreos.com/v1"
```
Comment (everettraven) [inline changelog/fragments/01-olm-install-retry.yaml:29]: Nit: I would recommend only showing the full error message to keep the changelog entry as short as possible but still informative

**Review by everettraven — APPROVED** [review]
Overall looks good to me. A couple nits:

**Comment by everettraven** [comment]
@nunnatsa it looks like the sanity check is failing due to an out of date fork. I think you need to fetch the latest changes on master and rebase this PR on top of that.

**Comment by nunnatsa** [comment]
> @nunnatsa it looks like the sanity check is failing due to an out of date fork. I think you need to fetch the latest changes on master and rebase this PR on top of that.

I think the sanity check is broken. I rebased my pr, again, just before thr lastest push.

File: internal/olm/client/client.go (hunk 5 above)
Comment (nunnatsa) [reply internal/olm/client/client.go:151]: Good point

**Comment by nunnatsa** [comment]
@everettraven - done

**Review by oceanc80 — APPROVED** [review]

**Review by rashmigottipati — APPROVED** [review]
/lgtm

**Review by everettraven — APPROVED** [review]
